### PR TITLE
Improve require speed

### LINKED ISSUE #180: Improve require speed
[Issuehunt badges](https://issuehunt.io/r/chalk/chalk/issues/180)

I tested this by creating a new project and adding just [time-require](https://www.npmjs.com/package/time-require) and chalk.  I then created a simple `time.js` file with the following contents:

```js
require('time-require')
require('chalk')
```

**Results for Chalk 1.x (1.1.3 as of this posting):**
```
Start time: (2017-07-22 16:50:03 UTC) [treshold=1%]
#  module                                                                time  %
1  pretty-ms (node_modules/pretty-ms/index.js)                            1ms  ▇▇▇▇▇▇▇▇ 11%
2  ansi-styles (node_modules/time-re...ules/ansi-styles/ansi-styles.js)   1ms  ▇▇▇▇▇▇▇▇ 11%
3  has-color (node_modules/has-color/index.js)                            1ms  ▇▇▇▇▇▇▇▇ 11%
4  chalk (node_modules/time-require/node_modules/chalk/index.js)          3ms  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 33%
5  escape-string-regexp (node_modules/escape-string-regexp/index.js)      1ms  ▇▇▇▇▇▇▇▇ 11%
6  ansi-styles (node_modules/ansi-styles/index.js)                        1ms  ▇▇▇▇▇▇▇▇ 11%
7  has-ansi (node_modules/has-ansi/index.js)                              1ms  ▇▇▇▇▇▇▇▇ 11%
8  chalk (node_modules/chalk/index.js)                                    3ms  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 33%
Total require(): 16
Total time: 9ms
```

**Results for Chalk 2.x (2.0.1 as of this posting):**
```Start time: (2017-07-22 16:48:22 UTC) [treshold=1%]
 #  module                                                                time  %
 1  text-table (node_modules/text-table/index.js)                          1ms  ▇▇▇▇▇ 6%
 2  date-time (node_modules/date-time/index.js)                            1ms  ▇▇▇▇▇ 6%
 3  parse-ms (node_modules/parse-ms/index.js)                              1ms  ▇▇▇▇▇ 6%
 4  pretty-ms (node_modules/pretty-ms/index.js)                            1ms  ▇▇▇▇▇ 6%
 5  ansi-styles (node_modules/time-re...ules/ansi-styles/ansi-styles.js)   1ms  ▇▇▇▇▇ 6%
 6  has-color (node_modules/has-color/index.js)                            1ms  ▇▇▇▇▇ 6%
 7  chalk (node_modules/time-require/node_modules/chalk/index.js)          2ms  ▇▇▇▇▇▇▇▇▇ 13%
 8  escape-string-regexp (node_modules/escape-string-regexp/index.js)      1ms  ▇▇▇▇▇ 6%
 9  ./conversions (node_modules/color-convert/conversions.js)              2ms  ▇▇▇▇▇▇▇▇▇ 13%
10  color-convert (node_modules/color-convert/index.js)                    5ms  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 31%
11  ansi-styles (node_modules/ansi-styles/index.js)                        6ms  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 38%
12  os (os)                                                                1ms  ▇▇▇▇▇ 6%
13  supports-color (node_modules/supports-color/index.js)                  1ms  ▇▇▇▇▇ 6%
14  ./templates.js (node_modules/chalk/templates.js)                       1ms  ▇▇▇▇▇ 6%
15  chalk (node_modules/chalk/index.js)                                   10ms  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 63%
Total require(): 20
Total time: 16ms
```

I ran the tests several times locally and the results were all within 1ms each time.

For long running processes, the difference here is obviously negligible so I'll understand if this gets closed out.

That being said, for CLI utils where boot time and responsiveness are far more important, it's unfortunate that just using 2.x adds an additional 5ms+ on average that you can't get back.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $60.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($60.00)
### Submitted pull Requests
- [#358 Improve require speed](https://issuehunt.io/r/chalk/chalk/pull/358)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/chalk/chalk/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by stroncium**
- Lazy load templates(runtime overhead should be zero due to prototype check overlapping with `=== undefined` check)
 - Don't access `ansi-styles` color space converters unless needed

This patch works in conjunction with(but works without them too, though to much smaller effect):
 - https://github.com/chalk/ansi-styles/pull/53
 - https://github.com/chalk/supports-color/pull/101

Require speed test results before:
before

Require speed test results after(pure require, without using template nor color conversion functionality):
after

Note: in both cases, a huge part of time is spent initializing `tty`(or initializing `process.stdout` and `process.stderr`, which includes initializing `tty`), which is also triggered by a lot of other parts of stdlib(for example by `console.log` with non-string arguments), and is required in order to detect if output supports color, so is expected to be initialized anyway by a lot of real life programs(notable exception: programs which conditionally produce no output). If we subtract the time required for this initializations, the final numbers will be ~10ms and ~5ms, respectively.

Fixes #180

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#180: Improve require speed](https://issuehunt.io/repos/11855195/issues/180)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

File: source/index.js [hunk 1]
```diff
@@ -76,11 +72,9 @@ styles.visible = {
 	}
 };
 
-for (const model of Object.keys(ansiStyles.color.ansi)) {
-	if (skipModels.has(model)) {
-		continue;
-	}
+const usedModels = ['rgb', 'hsl', 'hsv', 'hwb', 'cmyk', 'xyz', 'lab', 'lch', 'hex', 'keyword', 'ansi', 'ansi256', 'hcg', 'apple'];
```
Comment (sindresorhus) [inline source/index.js]: There are models here we don't officially support. Only include the list in the docs: https://github.com/chalk/chalk#256-and-truecolor-color-support

File: source/index.js [hunk 2]
```diff
@@ -15,9 +14,6 @@ const levelMapping = [
 	'ansi16m'
 ];
 
-// `color-convert` models to exclude from the Chalk API due to conflicts and such
-const skipModels = new Set(['gray']);
```
Comment (sindresorhus) [inline source/index.js]: Why was this removed?

File: source/index.js (hunk 1 above)
Comment (ExE-Boss) [reply source/index.js]:
```suggestion
const usedModels = ['rgb', 'hex', 'keyword', 'hsl', 'hsv', 'hwb', 'ansi', 'ansi256'];
```

File: source/index.js (hunk 2 above)
Comment (ExE-Boss) [reply source/index.js]: Probably because it’s no longer being used.

File: source/index.js (hunk 1 above)
Comment (sindresorhus) [reply source/index.js]: > Note that applying this will probably cause these models to become undefined:

Which is fine as they're not documented.

Comment (Qix-) [reply source/index.js]: Just noticed HCG should be supported (in a separate PR, though). Opened #360.

File: source/index.js (hunk 2 above)
Comment (stroncium) [reply source/index.js]: Yeah, no need for blacklist if we're using whitelist.

**Comment by stroncium** [comment]
With newer version of nodejs(I tested on 12.10.0) tty require speed increased a lot, and I'm currently getting 6.4 ms total require time with 3.8 going to `tty` on the same machine(but too lazy to relink all the stuff to compare with results prior to all the fixes here and in deps).

With this in mind, I think inlining basic color converters in this module to speed up require time even further doesn't really make sense anymore, as the gains should be minor. So I consider the patch complete in terms of functionality.

**Comment by ExE-Boss** [comment]
Since we’re removing undocumented colour spaces, this is probably a breaking change.

**Comment by stroncium** [comment]
@ExE-Boss changes in undocumented features is not considered API change, and not a breaking one for sure.

**Comment by sindresorhus** [comment]
@stroncium Can you fix the merge conflict?

**Comment by sindresorhus** [comment]
Awesome. Thanks for working on this 🙌

**Comment by Turbo87** [comment]
@stroncium out of curiosity: what tool did you use to create that terminal screenshot in the PR description?

**Comment by stroncium** [comment]
@Turbo87 That would be my private tool for node performance metrics and optimization. Didn't get to write docs for it and release publicly yet, sorry.
